official/coin_build/man/man3/SoCache.3
.TH "SoCache" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoCache \- The \fBSoCache\fP class is the superclass for all internal cache classes\&.
.PP
It organizes reference counting to make it possible to share cache instances\&. It also organizes a list of elements that will affect the cache\&. If any of the elements have changed since the cache was created, the cache is invalid\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/caches/SoCache\&.h>\fP
.PP
Inherited by \fBSoBoundingBoxCache\fP, \fBSoConvexDataCache\fP, \fBSoGLRenderCache\fP, \fBSoNormalCache\fP, \fBSoPrimitiveVertexCache\fP, and \fBSoTextureCoordinateCache\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSoCache\fP (\fBSoState\fP *const state)"
.br
.ti -1c
.RI "void \fBref\fP (void)"
.br
.ti -1c
.RI "void \fBunref\fP (\fBSoState\fP *state=NULL)"
.br
.ti -1c
.RI "void \fBaddElement\fP (const \fBSoElement\fP *const elem)"
.br
.ti -1c
.RI "virtual void \fBaddCacheDependency\fP (const \fBSoState\fP *state, \fBSoCache\fP *cache)"
.br
.ti -1c
.RI "virtual SbBool \fBisValid\fP (const \fBSoState\fP *state) const"
.br
.ti -1c
.RI "const \fBSoElement\fP * \fBgetInvalidElement\fP (const \fBSoState\fP *const state) const"
.br
.ti -1c
.RI "void \fBinvalidate\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual void \fBdestroy\fP (\fBSoState\fP *state)"
.br
.ti -1c
.RI "virtual \fB~SoCache\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSoCache\fP class is the superclass for all internal cache classes\&.
.PP
It organizes reference counting to make it possible to share cache instances\&. It also organizes a list of elements that will affect the cache\&. If any of the elements have changed since the cache was created, the cache is invalid\&. 

The cache element test algorithm in Coin works like this:
.PP
Every element that is read before it's written when a cache is created is stored in the \fBSoCache\fP's element list\&. This is done to detect when something outside the cache changes\&.
.PP
Example: you have a \fBSoCoordinate3\fP node outside an \fBSoSeparator\fP, but an \fBSoIndexedFaceSet\fP inside the \fBSoSeparator\fP\&. If the \fBSoSeparator\fP creates a cache, \fBSoIndexedFaceSet\fP will read \fBSoCoordinateElement\fP, and since \fBSoCoordinateElement\fP hasn't been set after the cache was opened, the cache stores that element in the cache's list of element dependencies\&.
.PP
At the next frame, the \fBSoSeparator\fP will test if the cache is valid, and will then test all dependency elements\&. If one of the elements doesn't match, the cache is not valid and can't be used\&.
.PP
That's the basics\&. There are some steps you have to do when creating a cache to make the cache dependencies work\&. Basically you have to do it like this:
.PP
.PP
.nf
SbBool storedinvalid = SoCacheElement::setInvalid(FALSE);
state->push();
SoMyCache * cache = new SoMyCache(state);
cache->ref();
SoCacheElement::set(state, cache);
buildMyCache();
state->pop();
SoCacheElement::setInvalid(storedinvalid);
.fi
.PP
.PP
First you reset and store the old value of the cache invalid-flag\&. Then you push the state so that the cache can detect when something outside the cache is changed (and to be able to change the cache element)\&. Next, you create the cache - don't forget to ref it\&. Finally, set the current cache in the cache element and build the cache\&. After building the cache, you pop the state and restore the invalid-cache flag\&.
.PP
When building the cache, all elements that are read will be copied into the cache (using \fBSoElement::copyMatchInfo()\fP), and these copied elements are used to test the validity of the cache (in \fBSoCache::isValid()\fP)\&.
.PP
You don't have to manually add element dependencies\&. They will automatically be picked up when creating the cache\&. This is handled in \fBSoElement::getConstElement()\fP\&.
.PP
If you want the cache to be invalidated when some field inside your node is changed, it's common to overload the notify()-method, and call \fBSoCache::invalidate()\fP whenever the notify()-method for your node is called\&. See for instance \fBSoShape::notify()\fP\&.
.PP
Also, don't delete the cache in your notify() method\&. Wait until the next time the cache is needed before unref-ing the old cache\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoCache::SoCache (\fBSoState\fP *const state)"
Constructor with \fIstate\fP being the current state\&. 
.SS "SoCache::~SoCache ()\fC [protected]\fP, \fC [virtual]\fP"
Destructor 
.SH "Member Function Documentation"
.PP 
.SS "void SoCache::ref (void)"
Increases the reference count by one\&. 
.SS "void SoCache::unref (\fBSoState\fP * state = \fCNULL\fP)"
Decreases the reference count by one\&. When the reference count reaches zero, the cache is deleted\&. The \fBSoCache::destroy()\fP method is called before the destructor is called\&. 
.SS "void SoCache::addElement (const \fBSoElement\fP *const elem)"
Adds \fIelem\fP to the list of elements this cache depends on\&. 
.SS "void SoCache::addCacheDependency (const \fBSoState\fP * state, \fBSoCache\fP * cache)\fC [virtual]\fP"
Adds dependencies from \fIcache\fP to this cache\&. 
.SS "SbBool SoCache::isValid (const \fBSoState\fP * state) const\fC [virtual]\fP"
Return \fITRUE\fP if this cache is valid, \fIFALSE\fP otherwise\&. 
.PP
Reimplemented in \fBSoPrimitiveVertexCache\fP, and \fBSoGLRenderCache\fP\&.
.SS "const \fBSoElement\fP * SoCache::getInvalidElement (const \fBSoState\fP *const state) const"
Returns the element that caused the invalidation\&. Returns \fINULL\fP if the cache is valid, or if the cache was not invalidated bacause of an element\&. 
.SS "void SoCache::invalidate (void)"
Forces a cache to be invalid\&. 
.SS "void SoCache::destroy (\fBSoState\fP * state)\fC [protected]\fP, \fC [virtual]\fP"
Can be overridden by subclasses to clean up before they are deleted\&. Default method does nothing\&. 
.PP
Reimplemented in \fBSoGLRenderCache\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/caches/SoCache.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \page caches The Cache Classes

  The cache classes are mostly internal to Coin.

  \ingroup caches
*/

/*!
  \class SoCache Inventor/caches/SoCache.h
  \brief The SoCache class is the superclass for all internal cache classes.
  \ingroup caches

  It organizes reference counting to make it possible to share cache
  instances. It also organizes a list of elements that will affect the
  cache. If any of the elements have changed since the cache was
  created, the cache is invalid.

  The cache element test algorithm in Coin works like this:

  Every element that is read before it's written when a cache is
  created is stored in the SoCache's element list. This is done to
  detect when something outside the cache changes.

  Example: you have a SoCoordinate3 node outside an SoSeparator, but
  an SoIndexedFaceSet inside the SoSeparator. If the SoSeparator
  creates a cache, SoIndexedFaceSet will read SoCoordinateElement, and
  since SoCoordinateElement hasn't been set after the cache was
  opened, the cache stores that element in the cache's list of element
  dependencies.

  At the next frame, the SoSeparator will test if the cache is valid,
  and will then test all dependency elements. If one of the elements
  doesn't match, the cache is not valid and can't be used.

  That's the basics. There are some steps you have to do when creating
  a cache to make the cache dependencies work. Basically you have to
  do it like this:
  
  \verbatim
  bool storedinvalid = SoCacheElement::setInvalid(false);
  state->push();
  SoMyCache * cache = new SoMyCache(state);
  cache->ref();
  SoCacheElement::set(state, cache);
  buildMyCache();
  state->pop();
  SoCacheElement::setInvalid(storedinvalid);
  \endverbatim
  
  First you reset and store the old value of the cache
  invalid-flag. Then you push the state so that the cache can detect
  when something outside the cache is changed (and to be able to
  change the cache element).  Next, you create the cache - don't
  forget to ref it. Finally, set the current cache in the cache
  element and build the cache. After building the cache, you pop the
  state and restore the invalid-cache flag.
  
  When building the cache, all elements that are read will be copied
  into the cache (using SoElement::copyMatchInfo()), and these
  copied elements are used to test the validity of the cache
  (in SoCache::isValid()).
  
  You don't have to manually add element dependencies. They will
  automatically be picked up when creating the cache. This is
  handled in SoElement::getConstElement().
  
  If you want the cache to be invalidated when some field inside your
  node is changed, it's common to overload the notify()-method, and
  call SoCache::invalidate() whenever the notify()-method for your
  node is called. See for instance SoShape::notify().

  Also, don't delete the cache in your notify() method. Wait until the
  next time the cache is needed before unref-ing the old cache.
*/


// FIXME: this really needs a usage example, preferably with source
// code for when using an extension cache. 20040722 mortene.

// *************************************************************************

#include <Inventor/caches/SoCache.h>

#include <cstring>
#include <cassert>

#include <Inventor/SbName.h>
#include <Inventor/elements/SoElement.h>
#include <Inventor/errors/SoDebugError.h>
#include <Inventor/lists/SbList.h>
#include <Inventor/misc/SoState.h>
#include <Inventor/C/tidbits.h>

#include "tidbitsp.h"
#include "coindefs.h"

// *************************************************************************

class SoCacheP {
public:
  SbList <SoElement *> elements;
  unsigned char * elementflags;
  int refcount;
  bool invalidated;
  int statedepth;
};

#define PRIVATE(obj) ((obj)->pimpl)

// *************************************************************************

/*!
  Constructor with \a state being the current state.
*/
SoCache::SoCache(SoState * const state)
{
  PRIVATE(this) = new SoCacheP;
  PRIVATE(this)->elementflags = nullptr;
  PRIVATE(this)->refcount = 0;
  PRIVATE(this)->invalidated = false;
  PRIVATE(this)->statedepth = state ? state->getDepth() : 0;

  int numidx = SoElement::getNumStackIndices();
  int numbytes = (numidx >> 3) + 1;
  // one bit per element is used to quickly determine whether an
  // element of a given type already has been added.
  PRIVATE(this)->elementflags = new unsigned char[numbytes];
  std::memset(PRIVATE(this)->elementflags, 0, numbytes);
}

/*!
  Destructor
*/
SoCache::~SoCache()
{
  delete [] PRIVATE(this)->elementflags;

  int n = PRIVATE(this)->elements.getLength();
  for (int i = 0; i < n; i++) {
    delete PRIVATE(this)->elements[i];
  }
  delete PRIVATE(this);
}

// *************************************************************************

/*!
  Increases the reference count by one.
*/
void
SoCache::ref(void)
{
  PRIVATE(this)->refcount++;
}

/*!
  Decreases the reference count by one. When the reference count reaches
  zero, the cache is deleted. The SoCache::destroy() method is called
  before the destructor is called.
*/
void
SoCache::unref(SoState *state)
{
  assert(PRIVATE(this)->refcount > 0);
  if (--PRIVATE(this)->refcount == 0) {
    this->destroy(state);
    delete this;
  }
}

// *************************************************************************

/*!
  Adds \a elem to the list of elements this cache depends on.
*/
void
SoCache::addElement(const SoElement * const elem)
{
  if (elem->getDepth() < PRIVATE(this)->statedepth) {
    int idx = elem->getStackIndex();
    int flag = 0x1 << (idx & 0x7);
    idx >>= 3; // get byte number
    if (!(PRIVATE(this)->elementflags[idx] & flag)) {
#if COIN_DEBUG // debug
      if (coin_debug_caching_level() > 1) {
        SoDebugError::postInfo("SoCache::addElement",
                               "cache: %p, elem: %s", this,
                               elem->getTypeId().getName().getString());
      }
#endif // debug
      SoElement * copy = elem->copyMatchInfo();
      if (copy) PRIVATE(this)->elements.append(copy);
      PRIVATE(this)->elementflags[idx] |= flag;
    }
  }
}

/*!
  Adds dependencies from \a cache to this cache.
*/
void
SoCache::addCacheDependency(const SoState * state, SoCache * cache)
{
  if (cache == this) return;

  // local variables for speed
  int n = cache->pimpl->elements.getLength();
  const SoElement * const * ptr = cache->pimpl->elements.getArrayPtr();
  for (int i = 0; i < n; i++) {
    // use elements in state to get correct element depth
    this->addElement(state->getConstElement(ptr[i]->getStackIndex()));
  }
}

/*!
  Return \e true if this cache is valid, \e false otherwise.
*/
bool
SoCache::isValid(const SoState * state) const
{
  if (PRIVATE(this)->invalidated) return false;
  return this->getInvalidElement(state) == nullptr;
}

/*!
  Returns the element that caused the invalidation. Returns \e nullptr
  if the cache is valid, or if the cache was not invalidated
  bacause of an element.
*/
const SoElement *
SoCache::getInvalidElement(const SoState * const state) const
{
  if (PRIVATE(this)->invalidated) return nullptr;

  // use local variables for speed
  int n = PRIVATE(this)->elements.getLength();
  const SoElement * const * ptr = PRIVATE(this)->elements.getArrayPtr();
  const SoElement * elem;
  for (int i = 0; i < n; i++) {
    elem = ptr[i];
    if (!elem->matches(state->getConstElement(elem->getStackIndex()))) {
#if COIN_DEBUG
      if (coin_debug_caching_level() > 0) {
        SoDebugError::postInfo("SoCache::getInvalidElement",
                               "cache: %p, invalid element: %s", this,
                               elem->getTypeId().getName().getString());
      }
#endif // debug
      return elem;
    }
  }
  return nullptr;
}

/*!
  Forces a cache to be invalid.
*/
void
SoCache::invalidate(void)
{
  PRIVATE(this)->invalidated = true;
}

/*!
  Can be overridden by subclasses to clean up before they are
  deleted. Default method does nothing.
*/
void
SoCache::destroy(SoState *)
{
}

#undef PRIVATE

// *************************************************************************
